sling/nlp/kb/calendar.cc

```diff
@@ -0,0 +1,394 @@
+    }
+    if (bc) year_ = -year_;
+
+    // Parse day and month.
+    if (p < end && *p == '-') {
+      p++;
+      p = parse_number(p, end, &month_);
+      if (month_ != 0) precision_ = MONTH;
+      if (p < end && *p == '-') {
+        p++;
+        p = parse_number(p, end, &day_);
+        if (day_ != 0) precision_ = DAY;
+      }
+    }
+  }
+}
+
+void Calendar::Init(Store *store) {
+  // Get symbols.
+  store_ = store;
+  n_name_ = store->Lookup("name");
+
+  // Get calendar from store.
+  Frame cal(store, "/w/calendar");
+  if (!cal.valid()) return;
+
+  // Get weekdays.
+  for (const Slot &s : cal.GetFrame("/w/weekdays")) {
+    weekdays_[s.name.AsInt()] = s.value;
+  }
+
+  // Get months.
+  for (const Slot &s : cal.GetFrame("/w/months")) {
+    months_[s.name.AsInt()] = s.value;
+  }
+
+  // Get days.
+  for (const Slot &s : cal.GetFrame("/w/days")) {
+    days_[s.name.AsInt()] = s.value;
+  }
+
+  // Get years.
+  for (const Slot &s : cal.GetFrame("/w/years")) {
+    years_[s.name.AsInt()] = s.value;
+  }
+
+  // Get decades.
+  for (const Slot &s : cal.GetFrame("/w/decades")) {
+    decades_[s.name.AsInt()] = s.value;
+  }
+
+  // Get centuries.
+  for (const Slot &s : cal.GetFrame("/w/centuries")) {
+    centuries_[s.name.AsInt()] = s.value;
+  }
+
+  // Get millennia.
+  for (const Slot &s : cal.GetFrame("/w/millennia")) {
+    millennia_[s.name.AsInt()] = s.value;
+  }
+};
+
+string Calendar::DateAsString(const Date &date) const {
+  // Parse date.
+  Text year = YearName(date.year());
+
+  switch (date.precision()) {
+    case Date::NONE:
+      return "";
+
+    case Date::MILLENNIUM: {
+      Text millennium = MillenniumName(date.year());
+      if (!millennium.empty()) {
+        return millennium.str();
+      } else if (date.year() > 0) {
+        return StrCat((date.year() - 1) / 1000 + 1, ". millennium AD");
+      } else {
+        return StrCat(-((date.year() + 1) / 1000 - 1), ". millennium BC");
+      }
+    }
+
+    case Date::CENTURY: {
+      Text century = CenturyName(date.year());
+      if (!century.empty()) {
+        return century.str();
+      } else if (date.year() > 0) {
+        return StrCat((date.year() - 1) / 100 + 1, ". century AD");
+      } else {
+        return StrCat(-((date.year() + 1) / 100 - 1), ". century BC");
+      }
+    }
+
+    case Date::DECADE: {
+      Text decade = DecadeName(date.year());
+      if (!decade.empty()) {
+        return decade.str();
+      } else {
+        return StrCat(year, "s");
+      }
+    }
+
+    case Date::YEAR: {
+      if (!year.empty()) {
+        return year.str();
+      } else if (date.year() > 0) {
+        return StrCat(date.year());
+      } else {
+        return StrCat(-date.year(), " BC");
+      }
+    }
+
+    case Date::MONTH: {
+      Text month = MonthName(date.month());
+      if (!month.empty()) {
+        if (!year.empty()) {
+          return StrCat(month, " ", year);
+        } else {
+          return StrCat(month, " ", date.year());
```

Comment:
should we not have a special case for date.year() < 0 ?
Similarly below

Reply:
We don't support relative dates, so we cannot just have "March 6" without a year.